packages/runtime/container-runtime/src/dataStore.ts

```diff
@@ -202,3 +206,48 @@ class DataStore implements IDataStore {
 		});
 	}
 }
+
+export class DataStore2 implements IDataStore2 {
+	private _dataStore?: IDataStore;
+	private attachedRuntime: boolean = false;
+	public get context(): IFluidDataStoreContextDetached {
+		return this.fluidDatastoreContext;
+	}
+	constructor(
+		private readonly fluidDatastoreContext: LocalDetachedFluidDataStoreContext,
+		private readonly internalId: string,
+		private readonly runtime: ContainerRuntime,
+		private readonly datastores: DataStores,
+		private readonly logger: ITelemetryLoggerExt,
+	) {}
+
+	private get dataStore(): IDataStore {
+		if (this._dataStore === undefined) {
+			assert(this.attachedRuntime, "DataStore2 must be attached to runtime before use");
+			this._dataStore = channelToDataStore(
+				this.fluidDatastoreContext.channelValue,
+				this.internalId,
+				this.runtime,
+				this.datastores,
+				this.logger,
+			);
+		}
+		return this._dataStore;
+	}
+
+	public get entryPoint(): IFluidHandle<FluidObject> {
+		return this.dataStore.entryPoint;
+	}
+
+	public async trySetAlias(alias: string): Promise<AliasResult> {
+		return this.dataStore.trySetAlias(alias);
+	}
+
+	public async attachRuntime(
```

Comment:
@vladsud It's fine to move `attachRuntime` from `IFluidDataStoreContextDetached` to here right? From what I can tell, the sole purpose of that API was to be able to "realize" a detached data store and now that can be done here.

Reply:
I find it a bit odd that we mix responsibilities (see my other comments).
This object has both trySetAlias() & attachRuntime(). I'd assume there should be no possibility (from type system) to call trySetAlias() on a context that has no runtime yet. I.e. maybe attachRuntime() should return IDataStore (or something similar). But having 2 workflows that are targeted at two vastly different usage patterns (component itself, and component users) seems wrong to me - it's similar to what we had long time before, where data store creation was returning a channel interface (channel interface is for internal component's and runtime use only)